_drafts/firmware-update-architecture.md

```diff
@@ -0,0 +1,834 @@
+directly.
+
+Typically, I would write a Python script to do this but in a pinch you can use
+some shell commands in a Makefile. For example, given `app.elf` I can do the
+following to get a `bin` file with the CRC added:
+
+```shell
+# Create "app.data", a binary object without our image header
+$ arm-none-eabi-objcopy app.elf app.data -O binary --remove-section=.image_hdr
+# Compute the CRC and write it to a binary file
+$ crc32 app.data | xxd -r -p | xxd -p -c1 | tac | xxd -r -p > app.crc
+# Create "app.bin", the full firmware binary
+$ arm-none-eabi-objcopy app.elf app.bin -O binary
+# Inject the CRC into "app.bin"
+$ dd if=app.crc of=app.bin obs=1 seek=4 conv=notrunc
+```
+
+That's it! We've added the CRC to our `bin` file, the main complexity really is
+the Endian swap. We can now display the information in our firmware:
+
+```c
+// app.c
+printf("App STARTED - version %d.%d.%d (%s) - CRC 0x%lx\n",
+       image_hdr.version_major,
+       image_hdr.version_minor,
+       image_hdr.version_patch,
+       image_hdr.git_sha,
+       image_hdr.crc);
+```
+
+And as expected we get:
+
+```
+App STARTED - version 1.0.1 (e8e8d53) - CRC 0xeb4f7f52
+```
+
+### Loading Images
+
+Both the Bootloader and the Loader need to load firmware images and start them.
+This includes:
+1. Locating them
+2. Verifying them, and
+3. Starting them.
+
+The code may look something like this:
+
+```c
+// boot.c
+for (image_slot_t slot = IMAGE_SLOT_1; slot < IMAGE_NUM_SLOTS; ++slot) {
+    const image_hdr_t *hdr = image_get_header(slot);
+    if (hdr == NULL) {
+        continue;
+    }
+    if (image_validate(slot, hdr) != 0) {
+        continue;
+    }
+
+    printf("Booting slot %d\n", slot);
+    image_start(hdr);
+}
+
+// No image found, go into DFU mode
+```
+
+Here our approach to locating images is simple: we use hardcoded "slots" where
+we expect images. Slot 1 is at `0x8004000`, and slot 2 is at `0x8020000`. We
+simply iterate through the slots looking for valid headers.
+
+Remember that the first field of our image header is the "magic" value. In
+`image_get_header` we read the header and verify the magic value.
+
+```c
+// image.c
+const image_hdr_t *image_get_header(image_slot_t slot) {
+    const image_hdr_t *hdr = NULL;
+
+    switch (slot) {
+        case IMAGE_SLOT_1:
+            hdr = (const image_hdr_t *)&__slot1rom_start__;
+            break;
+        case IMAGE_SLOT_2:
+            hdr = (const image_hdr_t *)&__slot2rom_start__;
+            break;
+        default:
+            break;
+    }
+
+    if (hdr && hdr->image_magic == IMAGE_MAGIC) {
+        return hdr;
+    } else {
+        return NULL;
+    }
+}
+```
+
+To validate the image, we compute its CRC and compare it with the value written
+in the header. This will allow you to catch data corruption or bit errors, and
+save you some trouble down the line. Note that verifying CRC on boot is not
+appropriate for every applicaton, as it can slow down boot significantly.
```

Comment:
could also call out its not appropriate for security reasons either / allude to a future post on the how to do signing